icu4j/main/tests/core/src/com/ibm/icu/dev/test/impl/UnitsTest.java

```diff
@@ -0,0 +1,366 @@
+        assertEquals("1 - eps: measures[1] unit", MeasureUnit.INCH.getIdentifier(),
+                measures2.get(1).getUnit().getIdentifier());
+
+        // Testing precision with meter and light-year. 1e-16 light years is
+        // 0.946073 meters, and double precision can provide only ~15 decimal
+        // digits, so we don't expect to get anything less than 1 meter.
+
+        // An epsilon's nudge under one light-year: should give 1 ly, 0 m.
+        input = MeasureUnit.LIGHT_YEAR;
+        output = MeasureUnit.forIdentifier("light-year-and-meter");
+        final MeasureUnitImpl inputImpl3 = MeasureUnitImpl.forIdentifier(input.getIdentifier());
+        final MeasureUnitImpl outputImpl3 = MeasureUnitImpl.forIdentifier(output.getIdentifier());
+
+        // TODO: reusing converter results in a leak.
+        ComplexUnitsConverter converter3 = new ComplexUnitsConverter(inputImpl3, outputImpl3, rates);
+
+        // TODO: reusing measures results in a leak.
+        List<Measure> measures3 = converter3.convert(BigDecimal.valueOf(2.0).subtract(ComplexUnitsConverter.EPSILON));
+        assertEquals("measures length", 2, measures3.size());
+        assertEquals("light-year test: measures[0] value", BigDecimal.valueOf(2), measures3.get(0).getNumber());
+        assertEquals("light-year test: measures[0] unit", MeasureUnit.LIGHT_YEAR.getIdentifier(),
+                measures3.get(0).getUnit().getIdentifier());
+        assertEquals("light-year test: measures[1] value", BigDecimal.ZERO, measures3.get(1).getNumber());
+        assertEquals("light-year test: measures[1] unit", MeasureUnit.METER.getIdentifier(),
+                measures3.get(1).getUnit().getIdentifier());
+
+        // 1e-15 light years is 9.46073 meters (calculated using "bc" and the CLDR
+        // conversion factor). With double-precision maths, we get 10.5. In this
+        // case, we're off by almost 1 meter.
+        List<Measure> measures4 = converter3.convert(BigDecimal.valueOf(1.0 + 1e-15));
+        assertEquals("measures length", 2, measures4.size());
+        assertEquals("light-year test: measures[0] value", BigDecimal.ONE, measures4.get(0).getNumber());
+        assertEquals("light-year test: measures[0] unit", MeasureUnit.LIGHT_YEAR.getIdentifier(),
+                measures4.get(0).getUnit().getIdentifier());
+        assertTrue("light-year test: measures[1] value", compareTwoBigDecimal(BigDecimal.valueOf(10),
+                BigDecimal.valueOf(measures4.get(1).getNumber().doubleValue()),
+                BigDecimal.valueOf(1)));
+        assertEquals("light-year test: measures[1] unit", MeasureUnit.METER.getIdentifier(),
+                measures4.get(1).getUnit().getIdentifier());
+
+        // 2e-16 light years is 1.892146 meters. We consider this in the noise, and
+        // thus expect a 0. (This test fails when 2e-16 is increased to 4e-16.)
+        List<Measure> measures5 = converter3.convert(BigDecimal.valueOf(1.0 + 2e-17));
```

Comment:
FYI: C++ and the comment here talks about 2e-16 - but the unit test has 2e-17? Why is it 2e-17? (I missed this in my review, sorry. Busy fixing it now.)

1.0 + 2e-17, at double precision, is basically just 1.0. What makes fixing this a bit less straight-forward:
Java is using BigDecimal. Why? As in: if we're using something more precise than double, particularly arbitrary precision, we need to decide how precise we want the result to be.

I'll see what I can figure out as a good place to draw the line. (If I had plenty of time, I'd benchmark to see how much slower BigDecimal is, and perhaps then argue to just stick to double precision, for Java too.)